crates/ruff_python_syntax_errors/src/lib.rs

```diff
@@ -105,6 +119,35 @@ impl SyntaxChecker {
                 });
             }
         }
+
+        self.duplicate_type_parameter_name(stmt);
+    }
+
+    fn duplicate_type_parameter_name(&mut self, stmt: &ast::Stmt) {
+        let (Stmt::FunctionDef(ast::StmtFunctionDef { type_params, .. })
+        | Stmt::ClassDef(ast::StmtClassDef { type_params, .. })
+        | Stmt::TypeAlias(ast::StmtTypeAlias { type_params, .. })) = stmt
+        else {
+            return;
+        };
+
+        let Some(type_params) = type_params else {
+            return;
+        };
```

Comment:
Not saying that we shouldn't change the definition but you could have written this as
```suggestion
        let type_params = match stmt {
            Stmt::FunctionDef(ast::StmtFunctionDef {
                type_params: Some(type_params),
                ..
            })
            | Stmt::ClassDef(ast::StmtClassDef {
                type_params: Some(type_params),
                ..
            }) => &**type_params,
            Stmt::TypeAlias(ast::StmtTypeAlias {
                type_params: Some(type_params),
                ..
            }) => type_params,
            _ => return,
        };
```

Reply:
For sure, I'm happy with that too if we'd rather not change the generated code.